### DISCUSSION TITLE
How to play audio from a mp4 video?

### DISCUSSION THREAD

**Discussion by Akascape** [PyAV-Org/PyAV#1260] [1. Help]
I created a video player for tkinter using PIL and pyav, and the video is playing properly without any issue. But it is lacking audio.
I want to know how I can add audio playback by reading the audio frames and play it, maybe using sounddevice or pyaudio...
The source code of the video player is here: https://github.com/Akascape/tkVideoPlayer 
(Summary: I simply used the `frame = next(self._container.decode(video=0)).to_image()` method to read the frames and show them in a label image using pillow)

Here is the progress:

https://user-images.githubusercontent.com/89206401/228515652-abe137d0-6823-4c56-ba5c-8eb8292e0182.mp4

**Comment by seeones** [comment]
Here is an example of playing audio using PyAudio. How to control the synchronization of audio and video during playback needs to be controlled in the main program.

```
class AudioRender():
    def play(self, path):
        import av
        import pyaudio
        container: av.container.InputContainer | av.container.Container = av.open(path)
        audio_stream: av.audio.AudioFrame = container.streams.audio[0]

        samplerate = audio_stream.rate
        channels = audio_stream.channels

        p = pyaudio.PyAudio()
        stream = p.open(format=pyaudio.paFloat32,
                        channels=channels,
                        rate=samplerate,
                        output=True)

        for frame in container.decode(audio_stream):
            audio_data = frame.to_ndarray().astype('float32')
            interleaved_data = audio_data.T.flatten().tobytes()
            stream.write(interleaved_data)

        stream.stop_stream()
        stream.close()
        p.terminate()
```

**Reply by Akascape** [reply]
@seeones Thanks, it is working, but there is a problem with sync, when I decode using `next(self._container.decode(video=0, audio=0))`, it generates video and audio frames unevenly with some video files which causes stuttering.

Example:
This type of decoding works good with sync:
```
video_frame

audio_frame
audio_frame

video_frame

audio_frame
audio_frame

video_frame
...
``` 
But this decoding doesn't sync properly:
```
video_frame
video_frame
video_frame

audio_frame 
audio_frame
audio_frame
audio_frame
audio_frame
audio_frame

video_frame
video_frame
video_frame
...
```
**Is there any way to evenly decode both frame types?**

This is the exact code:
```python
frame = next(self._container.decode(video=0, audio=0)) # simultaneously decode both audio and video

if 'Video' in repr(frame):
         # show the video frame
         self._current_img = frame.to_image()
         self._frame_number += 1               
else:
        # play the audio frame
         audio_data = frame.to_ndarray().astype('float32')
         interleaved_data = audio_data.T.flatten().tobytes()
         audio_stream.write(interleaved_data)
```

**Reply by MeteorShower2004** [reply]
I had modified your fork of `tkvideoplayer` to integrete in my app several months ago. It benifits me a lot. You did a excellent work! 🫡

Then I come to `pyav` 's repo to find resolution on _playing `s16` audio generates a lot of noise_ and find this discussion.

After experimenting, I find the simplest way to get things work, that is:

```python
#resample this s16 format frame
resampler = av.AudioResampler('fltp')
frame: AudioFrame
result = resampler.resample(frame)
# you need to check the result
# it's a list of frames
# however in this situation there seems only one frame in the list
frame = result[0]
interleaved_data = ... # do this as the normal situation
```

I hope this will be helpful for someone who finds this discussion.

> **Is there any way to evenly decode both frame types?**

As you had found, these containers are decoded in pakcages, and in the lowlevel `Container.decode()` iterates over package and decode frames for each package, which causes those frame are not sorted by time exactly. So that needs some work after decoding.

As for me, I use split the decoding thread into two threads (even two processes, if you.are going to support 720p or higher*), one for video decoding, one for audio decoding. They are popped out accoding to `pts` time when I need to play them. That's much more complex when you need to control the decoding process. But that works.

* It's crazy to play 4k video in tkinter 🥲. I finnally end up using PySDL to take over the drawing process of canvas to play videos...
